File: backend/messaging.py
"""Emergency messaging logic with route-aware delivery."""

from __future__ import annotations

from datetime import datetime, timezone

MESSAGE_HISTORY = []


def _timestamp():
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def get_recent_messages(limit=20):
    return list(MESSAGE_HISTORY[-limit:])


def send_message(network, sender, destination, message, priority="NORMAL"):
    """Send a message using the best available route."""
    from backend.routing import find_best_route

    sender_ok = sender in network.nodes
    destination_ok = destination in network.nodes
    if not sender_ok:
        raise ValueError(f"Sender '{sender}' is invalid.")
    if not destination_ok:
        raise ValueError(f"Destination '{destination}' is invalid.")

    priority = str(priority).upper()
    valid_priorities = {"LOW", "NORMAL", "HIGH", "CRITICAL"}
    if priority not in valid_priorities:
        raise ValueError(f"Priority must be one of: {sorted(valid_priorities)}.")

    route_result = find_best_route(network, sender, destination)
    route = route_result.get("route", [])
    if route_result.get("route_status") == "ok":
        delivery_status = "delivered"
    else:
        delivery_status = "failed"

    message_record = {
        "id": f"MSG-{len(MESSAGE_HISTORY) + 1:04d}",
        "sender": sender,
        "destination": destination,
        "message": message,
        "priority": priority,
        "timestamp": _timestamp(),
        "route": route,
        "delivery_status": delivery_status,
    }

    MESSAGE_HISTORY.append(message_record)
    return message_record

File: backend/messaging.ts
import { MeshNetwork } from './network.js';
import { findBestRoute, RoutingMetric, RouteHop } from './routing.js';

export interface MessageRecord {
  id: string;
  sender: string;
  destination: string;
  message: string;
  priority: 'LOW' | 'NORMAL' | 'HIGH' | 'CRITICAL';
  timestamp: string;
  route: string[];
  hop_count: number;
  total_latency_ms: number;
  delivery_status: 'delivered' | 'failed' | string;
  route_details?: RouteHop[];
  failure_reason?: string;
}

export const MESSAGE_HISTORY: MessageRecord[] = [];

export function getRecentMessages(limit = 30): MessageRecord[] {
  return MESSAGE_HISTORY.slice(-limit).reverse();
}

export function sendMessage(
  network: MeshNetwork,
  sender: string,
  destination: string,
  message: string,
  priority = 'NORMAL',
  metric: RoutingMetric = 'composite'
): { record: MessageRecord; routeResult: ReturnType<typeof findBestRoute> } {
  if (!network.nodes.has(sender)) {
    throw new Error(`Sender '${sender}' is invalid.`);
  }
  if (!network.nodes.has(destination)) {
    throw new Error(`Destination '${destination}' is invalid.`);
  }

  const normPriority = (String(priority).toUpperCase()) as MessageRecord['priority'];
  const validPriorities = new Set(['LOW', 'NORMAL', 'HIGH', 'CRITICAL']);
  if (!validPriorities.has(normPriority)) {
    throw new Error(
      `Priority must be one of: ${Array.from(validPriorities).sort().join(', ')}.`
    );
  }

  const routeResult = findBestRoute(network, sender, destination, metric);
  const route = routeResult.route;
  const isDelivered = routeResult.route_status === 'ok' && route.length > 0;
  const delivery_status = isDelivered ? 'delivered' : 'failed';

  const messageRecord: MessageRecord = {
    id: `EMG-${String(MESSAGE_HISTORY.length + 1).padStart(4, '0')}`,
    sender,
    destination,
    message,
    priority: normPriority,
    timestamp: new Date().toISOString(),
    route,
    hop_count: routeResult.hop_count,
    total_latency_ms: routeResult.total_latency_ms,
    delivery_status,
    route_details: routeResult.hops,
    failure_reason: isDelivered ? undefined : routeResult.reason,
  };

  MESSAGE_HISTORY.push(messageRecord);
  return { record: messageRecord, routeResult };
}
